src/qibolab/result.py

```diff
@@ -5,97 +5,186 @@
 import numpy as np
 import numpy.typing as npt
 
-ExecRes = np.dtype([("i", np.float64), ("q", np.float64)])
+iq_typing = np.dtype([("i", np.float64), ("q", np.float64)])
 
 
 @dataclass
-class ExecutionResults:
-    """Data structure to deal with the output of :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`"""
-
-    array: npt.NDArray[ExecRes]
-    shots: Optional[npt.NDArray[np.uint32]] = None
-
-    @classmethod
-    def from_components(cls, is_, qs_, shots=None):
-        ar = np.empty(is_.shape, dtype=ExecRes)
-        ar["i"] = is_
-        ar["q"] = qs_
-        ar = np.rec.array(ar)
-        return cls(ar, shots)
-
-    @property
-    def i(self):
-        return self.array.i
-
-    @property
-    def q(self):
-        return self.array.q
-
-    def __add__(self, data):
-        assert len(data.i.shape) == len(data.q.shape)
-        if data.shots is not None:
-            assert len(data.i.shape) == len(data.shots.shape)
-        # concatenate on first dimension; if a scalar is passed, just append it
-        axis = 0 if len(data.i.shape) > 0 else None
-        i = np.append(self.i, data.i, axis=axis)
-        q = np.append(self.q, data.q, axis=axis)
-        if data.shots is not None:
-            shots = np.append(self.shots, data.shots, axis=axis)
-        else:
-            shots = None
-
-        new_execution_results = self.__class__.from_components(i, q, shots)
-
-        return new_execution_results
+class IntegratedResults:
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.INTEGRATION and AveragingMode.SINGLESHOT
+    """
+
+    def __init__(self, i: np.ndarray, q: np.ndarray, shots=None):
+        self.voltage: npt.NDArray[iq_typing] = (
+            np.recarray((i.shape[0] // shots, shots), dtype=iq_typing)
+            if shots
+            else np.recarray(i.shape, dtype=iq_typing)
+        )
+        self.voltage["i"] = i.reshape(i.shape[0] // shots, shots) if shots else i
+        try:
+            self.voltage["q"] = q.reshape(q.shape[0] // shots, shots) if shots else q
+        except ValueError:
+            raise ValueError(f"Given i has length {len(i)} which is different than q length {len(q)}.")
```

Comment:
I suspect this error message to be rather close to the original one from NumPy. Consider dropping this, and just let the original one bubble up (for the sake of simplicity).